### PR TITLE
Toppra::SolvePathParameterization can specify a solver.

### PR DESCRIPTION
**Pull request by hongkai-dai**
I would like to give the user more flexibility to choose the solver, rather than hard-code the solver within Drake.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20664)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@AlexandreAmice for feature review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20664#-:-Nlao5G16XBDqC8JULdI:bgvexs6)* status: LGTM missing from assignee AlexandreAmice, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

File: multibody/optimization/toppra.cc [hunk 1]
```diff
@@ -602,14 +602,20 @@ Toppra::ComputeForwardPass(double s_dot_0,
...
   // Clp appears to outperform Mosek for these optimizations.
   // TODO(mpetersen94) Add Gurobi in the correct ordering
-  auto solver = solvers::MakeFirstAvailableSolver(
-      {solvers::ClpSolver::id(), solvers::MosekSolver::id()});
+  std::unique_ptr<solvers::SolverInterface> solver;
+  if (solver_id.has_value()) {
+    solver = solvers::MakeSolver(solver_id.value());
+  } else {
+    solver = solvers::MakeFirstAvailableSolver(
+        {solvers::ClpSolver::id(), solvers::MosekSolver::id()});
```
Comment (sadraddini) [inline multibody/optimization/toppra.cc:617]: Maybe Clarabel has to be added here?

**Review by sadraddini — COMMENTED** [review]
I think the new arg should be unit tested.

**Review by AlexandreAmice — COMMENTED** [review]
Ditto with Sadra. The new arg needs to be tested

> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20664#-:-NlcmGh53TAxcty3--4I:b-yvsvhf)* status: 1 unresolved discussion, LGTM missing from assignee AlexandreAmice, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

___
*[`multibody/optimization/toppra.cc` line 617 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20664#gh-1426805507:-NlclbTt3zsO4VJ2Fxr_:boh4len) ([raw file](https://github.com/RobotLocomotion/drake/blob/39e53599bdb3ae06efc12e1bdf9f21d41187dff6/multibody/optimization/toppra.cc#L617)):*
<details><summary><i>Previously, sadraddini (Sadra Sadraddini) wrote…</i></summary><blockquote>

Maybe Clarabel has to be added here? 
</blockquote></details>

Since `ClpSolver` is always available (i believe) then even including Mosek here seems pointless. I would change this to just be `ChooseBestSolver` or choose one, always available solver.

**Review by sadraddini — COMMENTED** [review]
Is this intended to solve https://github.com/RobotLocomotion/drake/issues/20619? 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20664#-:-Nldvqtj11u7E9QugsIx:bqfx852)* status: 1 unresolved discussion, LGTM missing from assignee AlexandreAmice, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

___
*[`multibody/optimization/toppra.cc` line 617 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20664#gh-1426805507:-NldvPS_BcRAycK5pA2b:b-iz4h3) ([raw file](https://github.com/RobotLocomotion/drake/blob/39e53599bdb3ae06efc12e1bdf9f21d41187dff6/multibody/optimization/toppra.cc#L617)):*
<details><summary><i>Previously, AlexandreAmice (Alexandre Amice) wrote…</i></summary><blockquote>

Since `ClpSolver` is always available (i believe) then even including Mosek here seems pointless. I would change this to just be `ChooseBestSolver` or choose one, always available solver.
</blockquote></details>

Accuracy and speed also matter. That's why it is better to go with `ChooseBestSolver`.

**Review by AlexandreAmice — COMMENTED** [review]
I think on the C++ side you should also add a test that ensures that the correct solver is being called.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20664#-:-NlhmOBI-nsX79S1tduL:be6c3ow)* status: 1 unresolved discussion, LGTM missing from assignee AlexandreAmice, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20664)* status: 2 unresolved discussions, LGTM missing from assignee AlexandreAmice, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/20664#-NmvJcq86OgkSftewDMY:-NmvJcq86OgkSftewDMZ:b-vu6uti) (no related file):*
BTW When providing a function argument to choose a solver, for generality we should accept a `const SolverInterface* solver`, not a `std::optional<SolverId>`.

The SolverId can only be one of the well-known solvers inside of  Drake.  The `SolverInterface` can be a third-party implementation of the interface.

Additionally, a user who is calling a function more than once could pass the same solver object over and over again (when passed by pointer), leading to improved efficiency.  When passed by SolverId, the function needs to create and then discard a new solver object every time.

Here's the start of a patch:

```patch
--- a/multibody/optimization/toppra.h
+++ b/multibody/optimization/toppra.h
@@ -113,11 +113,11 @@ class Toppra {
    * The path parameterization has the same start time as the original path's
    * starting break.
    *
-   * @param solver_id If not std::nullopt, then we will invoke this solver for
+   * @param solver When non-null, then we will invoke this solver for
    * the optimization problem. Otherwise Drake will choose the solver.
    */
   std::optional<PiecewisePolynomial<double>> SolvePathParameterization(
-      const std::optional<solvers::SolverId>& solver_id = std::nullopt);
+      const solvers::SolverInterface* solver = nullptr);
 
   /**
    * Adds a velocity limit to all the degrees of freedom in the plant. The
--- a/multibody/optimization/toppra.cc
+++ b/multibody/optimization/toppra.cc
@@ -603,18 +603,18 @@ Toppra::ComputeForwardPass(double s_dot_0,
 }
 
 std::optional<PiecewisePolynomial<double>> Toppra::SolvePathParameterization(
-    const std::optional<solvers::SolverId>& solver_id) {
+    const solvers::SolverInterface* solver) {
   const double s_dot_0 = 0;
   const double s_dot_N = 0;
 
-  // Clp appears to outperform Mosek for these optimizations.
-  // TODO(mpetersen94) Add Gurobi in the correct ordering
-  std::unique_ptr<solvers::SolverInterface> solver;
-  if (solver_id.has_value()) {
-    solver = solvers::MakeSolver(solver_id.value());
-  } else {
-    solver = solvers::MakeFirstAvailableSolver(
+  // If the user did not provide a solver, then choose one now.
+  std::unique_ptr<solvers::SolverInterface> temp_solver;
+  if (solver == nullptr) {
+    // Clp appears to outperform Mosek for these optimizations.
+    // TODO(mpetersen94) Add Gurobi in the correct ordering
+    temp_solver = solvers::MakeFirstAvailableSolver(
         {solvers::ClpSolver::id(), solvers::MosekSolver::id()});
+    solver = temp_solver.get();
   }
 
   const std::optional<Eigen::Matrix2Xd> K =
```

___

**Comment by jwnimmer-tri** [comment]
Closing as abandoned. When the work unpauses, feel free to reopen it, or open a new PR.

**Pull request closed without merging**